Document That `f16` And `f128` Hardware Support is Limited

ultrabear opened the pull request:
This adds a small paragraph to the recently added f16 and f128 types explaining that hardware support may be limited, and that performance may suffer as a result of that.

I mainly wrote this because I felt it may be useful to express in some form; as a launchpoint for readers of the documentation if they have issues with performance.

I tried to word the documentation in a way that doesn't create false assumptions (that f16/f128 is too slow to use, for instance), removing the software implementation part could mislead people to thinking that f16/f128 is only available on some platforms, not all, so I believe it is important to keep in.\
"not all *major* platforms" is specifically said so as to not be redundant, because not all platforms implement many things, but the average rustacean is probably going to be using x86_64 or aarch64 derived ISA's, which is who this documentation is targeted towards.

I'm not sure of the best way to word the documentation, or if it should even be added, but I feel like it may be useful to have (potentially in a reworded way, I'm not very confident in the current wording and cannot decide if that is because it is too vague to be useful or too specific to be generally correct).

ultrabear committed:
doc note that f16 and f128 hardware support is limited

Nilstrieb commented:
cc @tgross35

WiSaGaN commented:
Would it be helpful to actually state which platform is guaranteed to *have* support, so that user can have more information for decision making?
The proposed paragraph seems too vague to be helpful.

tgross35 commented on library/core/src/primitive_docs.rs:
```diff
@@ -1082,6 +1082,9 @@ impl<T> (T,) {}
 /// bits. Please see [the documentation for [`prim@f32`] or [Wikipedia on
 /// half-precision values][wikipedia] for more information.
 ///
+/// Note that not all major platforms have hardware support for f16, in which case a
+/// software implementation will be used. This may be slower than expected.
+///
```
We can probably be more specific about what happens here, maybe:

```suggestion
/// Note that most major platforms will provide `f16` math support by converting to and from
/// an `f32`, which is usually fairly performant but will not be as fast as using `f32` directly.
```

tgross35 commented on library/core/src/primitive_docs.rs:
```diff
@@ -1181,6 +1184,9 @@ mod prim_f64 {}
 /// as many bits as `f64`. Please see [the documentation for [`prim@f32`] or [Wikipedia on
 /// quad-precision values][wikipedia] for more information.
 ///
+/// Note that not all major platforms have hardware support for f128, in which case a
+/// software implementation will be used. This may be slower than expected.
+///
```
The wording can be a bit stronger since performance really is not great.

```suggestion
/// Note that most major platforms do not have hardware support for `f128`, in which case a
/// software implementation will be used. This can be significantly slower than using `f64`.
///
```

tgross35 reviewed (changes requested):
Thanks for updating this! I meant to add something myself at some point but didn't get to it.

> Would it be helpful to actually state which platform is guaranteed to have support, so that user can have more information for decision making?
> The proposed paragraph seems too vague to be helpful.

I don't think we need to be too specific here since the exact story is pretty fluid, e.g. new simd instructions keep making `f16` better. If anything maybe mention that s390x, as well as some versions of PowerPC and RiscV, have hardware `f128` acceleration since that it a lot more performance-dependent.

https://hackmd.io/ax9YKCEeSymE4kciy-BcIA#fn15 is where I collected more specific hardware support

tgross35 commented:
@rustbot label +A-docs +F-f16_and_f128

ultrabear committed:
Tgross feedback tweaks

Co-authored-by: Trevor Gross <t.gross35@gmail.com>

Update library/core/src/primitive_docs.rs

Co-authored-by: Trevor Gross <t.gross35@gmail.com>

Update library/core/src/primitive_docs.rs

tgross35 commented:
Those were just rough ideas, you can adjust the exact wording as you see fit.

That being said, LGTM. Feel free to squash.

workingjubilee commented:
@tgross35 SPARC64 V (XII, at least) also has f128, though we don't support SPARC very well.

workingjubilee commented on library/core/src/primitive_docs.rs:
```diff
@@ -1082,6 +1082,8 @@ impl<T> (T,) {}
 /// bits. Please see [the documentation for [`prim@f32`] or [Wikipedia on
 /// half-precision values][wikipedia] for more information.
 ///
+/// Note that most major platforms will provide `f16` math support by converting to and from
+/// an `f32`, which is usually fairly performant but will not be as fast as using `f32` directly.
```
We can be more precise here: The biggest reason to use `f16`, above all else, is for space reasons, when an application doesn't need the full bits for its exponent and precision.

Essentially all new hardware nowadays has native half-precision floating-point instruction support. Even the Raspberry Pi 5. However, that is often less than the minimal ABI-supported levels, so often only half-to-single-to-half-precision floating-point conversion instructions (which do it faster than one might do manually) can be presumed.

So to enable it to go fast requires enabling target features, often.

workingjubilee reviewed (commented):
It feels deeply incorrect to say "most major platforms" with the Apple M1 in the same room, and with most hardware people actually using having the target features. This is trying to talk about, and thus effectively skirting around, "rustc is a coward and doesn't enable anything but the minimum target features, so on x86-64 Linux this goes back **20 years** of hardware advances".

Let us actually be precise.

ultrabear commented:
> It feels deeply incorrect to say "most major platforms" with the Apple M1 in the same room, and with most hardware people actually using having the target features. This is trying to talk about, and thus effectively skirting around, "rustc is a coward and doesn't enable anything but the minimum target features, so on x86-64 Linux this goes back **20 years** of hardware advances".
> 
> Let us actually be precise.

I originally wrote "not all major platforms have hardware support", but that was before the change to talk about f32 downcasting, and lots of things are still building for only mostly recent amd64 (like 5-10 years back now) when avx2 was the latest thing, and avx512 either did not exist or had catastrophic issues with the early generations, the wording can probably use some work if it wasn't clear; most users on amd64 should not be assumed to have avx512

ultrabear committed:
Make f16 and f128 docs clearer on platform support

Co-authored-by: Jubilee <46493976+workingjubilee@users.noreply.github.com>

Update library/core/src/primitive_docs.rs

Rewrite f16 and f128 hw support comments to match PR feedback

I wrote RISC-V allcaps in all cases, and wrote amd64 lowercase in all
cases, im not sure if either is the more correct way for either
platform, thats just how I normally write them, if theres a precedent
elsewhere it should probably be changed to match though.

Update library/core/src/primitive_docs.rs

Co-authored-by: Jubilee <46493976+workingjubilee@users.noreply.github.com>

Update library/core/src/primitive_docs.rs

Co-authored-by: Jubilee <46493976+workingjubilee@users.noreply.github.com>

Update library/core/src/primitive_docs.rs

ultrabear commented on library/core/src/primitive_docs.rs:
```diff
@@ -1181,6 +1183,9 @@ mod prim_f64 {}
 /// as many bits as `f64`. Please see [the documentation for [`prim@f32`] or [Wikipedia on
 /// quad-precision values][wikipedia] for more information.
 ///
+/// Note that most major platforms do not have hardware support for `f128`, in which case a
+/// software implementation will be used. This can be significantly slower than using `f64`.
```
So I'm currently working on rewriting the f128 section with this feedback, but I don't really know how to approach "you need target features for it today" without being wrong in the future, should we insert a timestamp of some form (rustc version, literal date), or accept that this may become a maintenance burden in exchange for expressing these concepts today?

ultrabear replied:
Update: I've decided to just go forwards with potential maintenance burden, since this PR has had that element from the start regardless

ultrabear replied:
> I think Jubilee was just suggesting say that even on targets that have hardware support, you need to enable relevant target features to make use of it - not necessarily to actually make a list here (though an example would be fine). Some target features are listed at https://doc.rust-lang.org/reference/attributes/codegen.html#the-target_feature-attribute so it is fine to point there, but even that isn't always up to date...

Oh yeah it was never on my mind to actually list them out, just the idea that "you need target features" will probably be outdated someday, but as mentioned its a burden that is carried by this documentation inherently: the real world will change; commenting on it is inherently unstable

tgross35 replied:
Oh, I wouldn't even have considered that a maintenance burden 😆

ultrabear replied:
I have decided to enumerate that riscv has support, amd64 does not, and aarch64 does not, because they fall into the categories of popular archs that most people will use, or being a non mainframe tier processor that has support (riscv), adding the mainframe ISAs that support f128 just feels a bit much, turning the short example list into a more comprehensive looking thing (which it is not intended to be)

workingjubilee commented:
@ultrabear Yes, I understand you used some vagueness to avoid recommending anything overly temporally sensitive. I most certainly think that it is preferred to be more vague than not at all! I just think that _some_ temporal sensitivity is inevitable when commenting on hardware support, so we can accept the hit of having to update the docs eventually someday, in exchange for giving the reader some useful information.

ultrabear commented:
Finally got back to rewriting the docs, I don't think they are perfect yet but I hope they are significantly improved over the first iterations. I used wikipedia as a source url which I don't think is great? if we would rather link directly to what wikipedia links to there (the riscv specification) it can be changed, I just frankly did not read the riscv specification myself, so I didn't want to include it

workingjubilee commented on library/core/src/primitive_docs.rs:
```diff
@@ -1184,12 +1187,15 @@ mod prim_f64 {}
 ///
-/// Note that most major platforms do not have hardware support for `f128`, in which case a
-/// software implementation will be used. This can be significantly slower than using `f64`.
+/// Note that no platforms have hardware support for `f128` without enabling target specific features
+/// (and [only some consumer level hardware has support][wikipedia-support], for example RISC-V has support, but
+/// neither amd64 nor aarch64 has support), in which case a software implementation will be used. This can be
+/// significantly slower than using `f64`.
```
Sorry, I think my prior comments may have been slightly confusing on this point. 

I am not aware of *hardware implementations* for RISCV. It is _specified in the ISA spec_, and it might be implemented in an FPGA soft-core somewhere, but "just run a custom FPGA core" is synonymous with "no hardware support" for most Rust programmers.

And the RISCV situation is similar to the SPARC64 situation: they specified it long before it was implemented.

Also, I checked and in PowerISA 3.1 they now specify compliancy subsets and quadruple-precision is also optional in most subsets, so yeah, the "base ISA" doesn't include it.

My thought with "major platforms" was that it requires a very subjective definition of major. Basically, I don't think we should try to start arguments with IBM or SiFive. But yes, realistically, people broadly don't have this. I think the important thing to capture here is to inform people that *even when an ISA says it has so-and-so*, only very specific *implementations of that ISA* will implement that *optional extension*.

workingjubilee commented on library/core/src/primitive_docs.rs:
```diff
@@ -1082,9 +1082,12 @@ impl<T> (T,) {}
-
+/// Note that most common platforms will not support `f16` in hardware without enabling extra target
+/// features, with the notable exception of Apple Silicon (also known as M1/2/etc) processors. For
+/// amd64 this requires the avx512fp16 feature, while RISC-V requires Zhf or Zhfmin, to have hardware support.
```
We should probably say x86-64 here? And we may wish to quote the string if we're using the Rust feature name specifically. Also Zhfmin only implements conversions, so,

```suggestion
/// Hardware support on x86-64 requires the avx512fp16 feature, while RISC-V requires Zhf.
```

Alternately we can give a small table?

```suggestion
/// | arch | feature | supported instructions |
/// | --- | --- | --- |
/// | x86_64 | f16c | f16 <-> f32 conversions |
/// | x86_64 | avx512fp16 | f16 arithmetic |
/// | riscv | Zhfmin | f16 <-> float conversions |
/// | riscv | Zhf | f16 arithmetic |
```

ultrabear replied:
> We should probably say x86-64 here? And we may wish to quote the string if we're using the Rust feature name specifically. Also Zhfmin only implements conversions, so,
> 
> Alternately we can give a small table?

well heres the issue with quoting off wikipedia lol (Zhfmin), I should probably have read the spec, apologies there

workingjubilee replied:
yeah, I'm not wild about the table idea, just thought about it.

workingjubilee commented on library/core/src/primitive_docs.rs:
```diff
@@ -1082,9 +1082,12 @@ impl<T> (T,) {}
-/// an `f32`, which is usually fairly performant but will not be as fast as using `f32` directly.
-
+/// Note that most common platforms will not support `f16` in hardware without enabling extra target
+/// features, with the notable exception of Apple Silicon (also known as M1/2/etc) processors. For
```
We can spare a few bytes.
```suggestion
/// features, with the notable exception of Apple Silicon (also known as M1, M2, etc.) processors.
```

workingjubilee reviewed (commented):
Nice! The f16 doc addition looks much better now. I think I may have wandered into the weeds a bit on my remarks re: f128, sorry about that!

workingjubilee commented on library/core/src/primitive_docs.rs:
```diff
@@ -1181,9 +1187,15 @@ mod prim_f64 {}
 /// quad-precision values][wikipedia] for more information.
 ///
+/// Note that no platforms have hardware support for `f128` without enabling target specific features
+/// (and [only some consumer level hardware has support][wikipedia-support], for example RISC-V has support, but
+/// neither amd64 nor aarch64 has support), in which case a software implementation will be used. This can be
+/// significantly slower than using `f64`.
```
```suggestion
/// as for all instruction set architectures, `f128` is considered an optional feature.
/// Only Power ISA ("PowerPC") and RISCV specify it, and only certain microarchitectures
/// actually implement it. For x86-64 and AArch64, ISA support is not even specified,
/// so it will always be a software implementation significantly slower than `f64`.
```

ultrabear committed:
Make f128 docs mention lack of any normal platform support

Co-authored-by: Jubilee <46493976+workingjubilee@users.noreply.github.com>

Update library/core/src/primitive_docs.rs

Remove orphaned doc link and clean up grammar a bit

Update library/core/src/primitive_docs.rs

workingjubilee commented:
Thanks!

@bors r+ rollup

bors commented:
:pushpin: Commit 5aa2f9a208162389a743c45cebb72a994effca1c has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 5aa2f9a208162389a743c45cebb72a994effca1c -->
<!-- homu: {"type":"Approved","sha":"5aa2f9a208162389a743c45cebb72a994effca1c","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 5aa2f9a208162389a743c45cebb72a994effca1c with merge 3fbeedb8c8d78feaefaa42613dcbac24fec04bf0...
<!-- homu: {"type":"BuildStarted","head_sha":"5aa2f9a208162389a743c45cebb72a994effca1c","merge_sha":"3fbeedb8c8d78feaefaa42613dcbac24fec04bf0"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8955778813/job/24596863686)
Approved by: workingjubilee
Pushing 3fbeedb8c8d78feaefaa42613dcbac24fec04bf0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"workingjubilee","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8955778813/job/24596863686"},"merge_sha":"3fbeedb8c8d78feaefaa42613dcbac24fec04bf0"} -->

Kobzol commented:
This PR has been merged by accident, without the full test suite being run. It was my fault (https://github.com/rust-lang/rust/pull/124631 got merged and it broke our CI). @ultrabear Could you please create a new PR with your branch against the latest version of `origin/master`? Thank you, and sorry for the mess.

dtolnay commented:
New PR, from the same branch: #124750
